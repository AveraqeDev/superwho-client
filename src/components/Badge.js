import React from 'react';

import '../styles/Badge.css';

const Badge = ({ values }) => {
  return (
    <>
      {values.split(',').map((value, idx) => {
        return (
          <span key={idx} className="badge">
            {value}
          </span>
        );
      })}
    </>
  )
};

export default Badge;