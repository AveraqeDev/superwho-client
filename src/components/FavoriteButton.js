import React from 'react';

import '../styles/FavoriteButton.css';

const FavoriteButton = ({ heroId }) => {

  const onClick = e => {
    e.preventDefault();
    e.currentTarget.classList.toggle('is_animating');

    e.currentTarget.classList.add('active');
  };

  const onAnimationEnd = e => {
    e.preventDefault();
    e.currentTarget.classList.toggle('is_animating');
  }
  return (
    <span 
      className={'favme'}
      onClick={onClick}
      onAnimationEnd={onAnimationEnd}
    >â™¥</span>
  );
};

export default FavoriteButton;