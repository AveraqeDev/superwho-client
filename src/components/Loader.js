import React from 'react';

import '../styles/Loader.css';

const Loader = () => {
  return (
    <div className="Loader">
      <div className="Loader__bounce1" />
      <div className="Loader__bounce2" />
      <div className="Loader__bounce3" />
    </div>
  );
}

export default Loader;