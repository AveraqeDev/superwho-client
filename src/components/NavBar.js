import React, { useState, useEffect, useContext } from 'react';
import { Link, withRouter } from 'react-router-dom';
import { CSSTransition } from 'react-transition-group';

import TokenService from '../services/token-service';
import { UserContext } from '../contexts/UserContext';

import '../styles/NavBar.css';

const Branding = React.memo(function Branding() {
  return (
    <Link className="Branding" to="/">SuperWho?</Link>
  );
});

const Burger = React.memo(function Burger({ open, setOpen }) {
  return (
    <button
    aria-label='Burger Menu'
    className='Burger'
    onClick={() => setOpen(!open)}
    >
      <div className={open ? 'burger_open' : ''} />
      <div className={open ? 'burger_open' : ''} />
      <div className={open ? 'burger_open' : ''} />
    </button>
  )
});

const Menu = React.memo(function Menu({ open, setOpen, onLogout, location }) {
  const { pathname } = location;

  const { user } = useContext(UserContext);

  return (
    <div className="Menu">
      <Link className={`Menu__Link${pathname === '/' ? ' Menu__Link-selected' : ''}`} to='/' onClick={() => setOpen(!open)}>Search</Link>
      {(user.id)
        ? <Link className={`Menu__Link${pathname === '/favorites' ? ' Menu__Link-selected' : ''}`} to='/favorites' onClick={() => setOpen(!open)}>Favorites</Link>
        : ''
      }
      {(user.id)
        ? <Link className={`Menu__Link${pathname === '/profile' ? ' Menu__Link-selected' : ''}`} to='/profile' onClick={() => setOpen(!open)}>Profile</Link>
        : <Link className={`Menu__Link${pathname === '/login' ? ' Menu__Link-selected' : ''}`} to='/login' onClick={() => setOpen(!open)}>Login</Link>
      }
      {(user.id)
        ? <button className="Menu__Link" onClick={onLogout}>Logout</button>
        : <Link className={`Menu__Link${pathname === '/register' ? ' Menu__Link-selected' : ''}`} to='/register' onClick={() => setOpen(!open)}>Register</Link>
      }
    </div>
  );
});

const UserAccount = React.memo(function UserAccount({ open, setOpen }) {
  const { user } = useContext(UserContext);

  if(user) {
    return (
      <div className="UserAccount">
        <p onClick={() => setOpen(!open)}>{user.username} <span className="DownArrow">â–¼</span></p>
      </div>
    );
  } else {
    return (
      <div className="UserAccount">
        <Link className="UserAccount__Link" to="/login">Login</Link>
        <Link className="UserAccount__Link" to="/register">Register</Link>
      </div>
    );
  }
});

const UserAccountMenu = React.memo(function UserAccountMenu({ open, setOpen, onLogout }) {
  return (
    <div className="UserAccountMenu">
      <Link className="UserAccountMenu__Link" to="/profile" onClick={() => setOpen(!open)}>Profile</Link>
      <hr className="seperator"></hr>
      <button className="UserAccountMenu__Link" onClick={onLogout}>Logout</button>
    </div>
  );
});

const Sidebar = React.memo(function Sidebar(props) {
  const { pathname } = props.location;

  const { user } = useContext(UserContext);

  return (
    <div className="Menu">
      <Link className={`Menu__Link${pathname === '/' ? ' Menu__Link-selected' : ''}`} to='/'>Search</Link>
      {(user)
        ? <Link className={`Menu__Link${pathname === '/favorites' ? ' Menu__Link-selected' : ''}`} to='/favorites'>Favorites</Link>
        : ''
      }
    </div>
  );
});

const NavBar = (props) => {
  const [navOpen, setNavOpen] = useState(false);
  const [accountOpen, setAccountOpen] = useState(false);
  const [windowWidth, setWindowWidth] = useState(0);
  const resizeWindow = () => {
    setWindowWidth(window.innerWidth);
  };

  const { setUser } = useContext(UserContext);

  useEffect(() => {
    resizeWindow();
    window.addEventListener("resize", resizeWindow);
    return () => window.removeEventListener("resize", resizeWindow);
  }, []);

  const onLogout = () => {
    setNavOpen(false);
    setAccountOpen(false);
    TokenService.clearAuthToken();
    setUser(null);
    props.history.push('/');
  };

  if(props.location.pathname === '/login' || props.location.pathname === '/register') {
    return '';
  }

  if(windowWidth <= 760) {
    return (
      <nav className="NavBar">
        <Burger open={navOpen} setOpen={setNavOpen} />
        <CSSTransition
          in={navOpen}
          timeout={300}
          classNames='Menu'
          unmountOnExit
        >
          <Menu open={navOpen} setOpen={setNavOpen} onLogout={onLogout} location={props.location} />
        </CSSTransition>
        <Branding />
      </nav>
    );
  } else {
    return (
      <nav className="NavBar">
        <Sidebar location={props.location} />
        <Branding />
        <UserAccount open={accountOpen} setOpen={setAccountOpen} />
        <CSSTransition
          in={accountOpen}
          timeout={300}
          classNames='AccountMenu'
          unmountOnExit
        >
          <UserAccountMenu open={accountOpen} setOpen={setAccountOpen} onLogout={onLogout} />
        </CSSTransition>
      </nav>
    )
  }
};

export default withRouter(NavBar);