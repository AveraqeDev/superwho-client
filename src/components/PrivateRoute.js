import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import { useAuthState } from "../hooks/AuthState";

const PrivateRoute = ({ children, ...props }) => {
  const { authenticated } = useAuthState()
  return (
    <Route 
      {...props}
    >
      {authenticated ? (
        {children}
      ) : (
        <Redirect
          to='/login'
        />
      )}
    </Route>
  );
};

export default PrivateRoute;