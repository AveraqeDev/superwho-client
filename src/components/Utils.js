import React from 'react';

import '../styles/Utils.css';

export const Input = React.memo(function Input({ className, ...props }) {
  return <input className={['Input', className].join(' ')} {...props} />
});

export const Button = React.memo(function Button({ className, ...props }) {
  return <button className={['Button', className].join(' ')} {...props} />
});