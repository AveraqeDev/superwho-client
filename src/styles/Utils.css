.Input {
  display: block;
  background: #EFEFEF;
  border: none;
  border-radius: 5px;
  box-shadow: inset 0px 4px 4px rgba(0, 0, 0, 0.25);
  color: #CBD0D6;
  margin-bottom: 10px;
  margin-left: auto;
  margin-right: auto;
  padding: 10px;
  text-align: left;
}

.Input::placeholder {
  color: #CBD0D6;
  font-size: 16px;
}

.Button {
  background: #2A9BE4;
  border: none;
  border-radius: 5px;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  color: #FFFFFF;
  cursor: pointer;
  font-weight: bold;
}