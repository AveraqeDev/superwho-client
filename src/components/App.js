import React, { useEffect } from 'react';
import { Route, Switch, useLocation } from 'react-router-dom';

import api from '../api'
import { useAuthState } from '../hooks/AuthState';
import PrivateRoute from '../components/PrivateRoute';
import PublicRoute from '../components/PublicRoute';

import NavBar from './NavBar';

import Login from '../pages/Login';
import Register from '../pages/Register';

import Search from '../pages/Search';
import Favorites from '../pages/Favorites';

const App = () => {
  const { key } = useLocation();
  const { authenticated, dispatch } = useAuthState();

  useEffect(() => {
    let isCurrent = true
    if(!authenticated) {
      api.auth.getAuthenticatedUser().then(user => {
        if(user && isCurrent) {
          dispatch({ type: 'LOGIN', user })
        }
      })
    }
    return () => (isCurrent = false)
  }, [authenticated, dispatch])

  useEffect(() => {
    window.scrollTo(0, 0)
  }, [key])

  return (
    <div className="App">
      <header>
        <NavBar />
      </header>
      <main>
          <Switch>
            <PublicRoute
              path='/login'
              component={Login}
            />
            <PublicRoute
              path='/register'
              component={Register}
            />

            <Route
              exact
              path='/'
              component={Search}
            />

            <PrivateRoute 
              path='/favorites'
              component={Favorites}
            />
          </Switch>
      </main>
    </div>
  );
}

export default App;
