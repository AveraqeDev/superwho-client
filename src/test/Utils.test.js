import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { Input, Button } from '../components/Utils';


describe('Utils', () => {
  describe('<Input />', () => {
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <Input 
          required
          name='test'
          id='test'
          placeholder='This is a test'
        />, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <Input 
            required
            name='test'
            id='test'
            placeholder='This is a test'
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<Button />', () => {
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<Button>Test Button</Button>, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <Button>Test Button</Button>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });
});