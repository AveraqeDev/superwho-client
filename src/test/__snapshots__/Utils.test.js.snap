// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils <Button /> renders the UI as expected 1`] = `
<button
  className="Button "
>
  Test Button
</button>
`;

exports[`Utils <Input /> renders the UI as expected 1`] = `
<input
  className="Input "
  id="test"
  name="test"
  placeholder="This is a test"
  required={true}
/>
`;
